design/bridge_design.md

```diff
@@ -0,0 +1,364 @@
+
+```solidity 
+function queryMessages(
+        address      sender,
+        bytes memory topic,
+        uint256      fromIndex,
+        uint256      toIndex,
+) public returns (bytes [] memory)
+```
+
+This function should return all the messages between the requested `fromIndex` and `toIndex` of the array storing the submitted messages. The array can be found when going through the map of `senders` under the concrete `topic`. It is a more expensive way to get messages on-chain without relying on users to ferry them around. 
+
+### MessageBus Internal workings - L1 to L2
+
+When a block from `L1` is processed by the `enclave` and transactions inside of it result in `MessagePublished` events, the `enclave` will submit them as messages to the inbox of the `L2` contract. This enables the `L2 smart contracts` to use/consume messages from `L1`. 
+
+![Diagram not found](./resources/PublishFromEthereum.svg)
+
+> **_NOTE:_** The enclave should not accept user transactions to system contracts. Only internal/synthetic transactions should be able to call the system functions!
+
+### MessageBus Internal workings - L2 to L1
+
+When a transaction on the `L2` results in `LogMessagePublished`, the event will automatically be added to the `Rollup header` by the `enclave`. Then the management contract will submit them to the `MessageBus` or they will directly be.
+
+![Diagram not found](./resources/PublishFromObscuro.svg)
+
+> **_NOTE:_** **The messages must not be accessible unless** the challenge period has passed! On top of that the block where the message is submitted to L1 must have confirmations equal to `consistencyLevel` before the message is released. Those are simply counted on-chain as "confirmations" is meaningless for Obscuro L2.  
+
+### Alternative approaches
+
+1. Obscuro only ever pushes the hash of the message. The user has the responsibility of providing the full message which will only be accepted if it matches one of the hashes, if neccessary.
+  * This simplifies gas cost calculations, but the problem described in the `Fees` section remains.
+  * Contracts can hash their messages before passing them to the `MessageBus` and achieve nearly the same outcome if they want to. 
+2. Obscuro only pushes to L2. Messages on L1 are provided signed by the enclave through an RPC and the MessageBus contract verifies that they have been signed by a correct enclave-owned key. 
+  * This disabled the option to read messages and ordering becomes problematic.
+  * The API is different between layers, which might end up being a bad UX.
+### Fees
+
+When publishing a message on the Obscuro L2, storing the message will have a direct cost to the `Sequencer` who is publishing the Rollup in the form of gas. In order to channel this cost to the user who is publishing the message we would need some configurable properties.
+
+**Assumption: Obscuro DAO** will vote and set the following properties:
```

Comment:
I'm partial to having designs with clear decisions, would be cool if we could make a decision and leave it documented, but I understand it might take a few more iterations